페이지네이션, entity 구조 개선 및 Swagger 반영

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈
- #20

## 📝 작업 내용
> **페이지네이션 구조 개선**
`PaginationRoomResponseDto` 추가로 API 응답 표준화
서비스 계층에서 util 기반 페이지네이션 로직 적용하고 컨트롤러에서 DTO를 활용한 응답 반환
**Swagger 문서화 도입**
Swagger 설치 및 기본 설정
기존 API 코드에 Swagger 데코레이터 적용
API 문서 확인 및 테스트 완료

yxxjxxeee committed:
feat: (#20) Swagger 설치 및 main.ts 설정 추가

yxxjxxeee committed:
Merge branch 'develop' of https://github.com/modern-agile-team/10term-pollycasso-back into feat/#20/init-swagger-docs

yxxjxxeee committed:
refactor: (#20) 페이지네이션 구조 개선 및 Swagger 반영

- PaginationRoomResponseDto 생성
- 서비스에서 util 기반으로 페이지네이션 처리
- 컨트롤러에서 PaginationRoomResponseDto 응답 사용
- Swagger 문서 작성

yxxjxxeee committed:
Merge branch 'develop' into feat/#20/init-swagger-docs

yxxjxxeee committed:
chore: (#20) package-lock.json 추가

yxxjxxeee committed:
chore: (#20) package-lock.json 업데이트

rrgks6221 commented on src/main.ts:
```diff
@@ -3,13 +3,16 @@ import { AppModule } from './app.module';
 
   app.enableCors({
-    origin: 'https://www.pollycasso.com',
+    origin: ['http://localhost:3000', 'https://api.pollycasso.com'],
```
https://api.pollycasso.com 는 백엔드 주소일거같은데 맞을까요??
프론트엔드 주소가 추가되야할거같아요

rrgks6221 commented on src/main.ts:
```diff
@@ -3,13 +3,16 @@ import { AppModule } from './app.module';
-    origin: 'https://www.pollycasso.com',
+    origin: ['http://localhost:3000', 'https://api.pollycasso.com'],
     credentials: true,
+    methods: 'GET,POST,PUT,PATCH,DELETE',
```
프론트엔드 http 클라이언트 라이브러리에서  HEAD 요청할 수 있어서 있어서 여쭤보시고 필요시에 추가 돼야할 수도 있을거같아요

rrgks6221 commented on src/main.ts:
```diff
@@ -36,8 +39,36 @@ async function bootstrap() {
+  const swaggerDocument = SwaggerModule.createDocument(app, swaggerConfig);
+  SwaggerModule.setup('api', app, swaggerDocument, {
+    swaggerOptions: {
+      persistAuthorization: true,
```
공식문서에 없는 내용이라 찾기 어려웠을텐데 잘 적용하셨네요

rrgks6221 commented on src/rooms/dtos/requests/create-room.dto.ts:
```diff
@@ -33,10 +48,19 @@ export class CreateRoomDto {
   isPrivate: boolean;
 
   @IsOptional()
   @IsNumberString({ no_symbols: true })
   @Length(4, 4)
+  @ApiProperty({
+    description: '비공개 방 비밀번호 (4자리 숫자)',
+    example: '1234',
+    required: false,
+  })
```
문서 스펙을 통해 4자리만 허용함을 알려주고자한다면 아래처럼 표현할 수도 있을거같아요

```suggestion
  @ApiProperty({
    description: '비공개 방 비밀번호 (4자리 숫자)',
    example: '1234',
    required: false,
    minLength: 4,
    maxLength: 4,
  })
```

rrgks6221 commented on src/rooms/dtos/requests/query-room.dto.ts:
```diff
@@ -18,14 +28,28 @@ export class QueryRoomDto {
   @IsOptional()
   @Type(() => Number)
   @IsInt()
+  @Min(1)
+  @ApiProperty({
+    description: '페이지네이션 커서 (마지막으로 불러온 방 ID)',
+    required: false,
+  })
```
개인적으로 request validation을 한다면 문서에 포함하는것이 좋다고 생각하기 때문에 minLength를 추가하는건 어떻게 생각하실까요??
문서에 없는데 에러가 발생하는 경우 한번씩 문의가 있을 수 있어요. 관리하는 입장에서는 귀찮은 일이기는 하지만요

```suggestion
  @ApiProperty({
    description: '페이지네이션 커서 (마지막으로 불러온 방 ID)',
    required: false,
    minLength: 1,
  })
```

rrgks6221 commented on src/rooms/dtos/requests/update-room.dto.ts:
swagger문서의 표현과 벨리데이터 데코레이터를 봤을 때 모든 필드가 optional 필드인데 타입은 전부 required 타입으로 되어있어서 수정돼야할거같아요

rrgks6221 commented on src/rooms/dtos/requests/update-room.dto.ts:
```diff
@@ -1,4 +1,75 @@
+export class UpdateRoomDto {
+  @IsString()
+  @IsNotEmpty()
+  @MaxLength(15)
```
아래 순서로 데코레이터를 배치한 이유가 string 이어야 empty 검사는게 의미가 있고 empty 검사가 돼야 maxLength를 검사하는 의미가 있기 때문에 이러한 배치를 한것일까요??

1. IsString
2. ISNotEmpty
3. MaxLength

yxxjxxeee replied:
맞습니다. 위 순서대로 배치해야 순서가 논리적이라서 이렇게 배치했습니다.

rrgks6221 replied:
그런 이유라면 역순으로 배치하는게 좋을거같습니다. 현재 validation pipe 설정에서는 유의미한 차이가 없음

이전에 9기 분에게 남겼던 코멘트예요 https://github.com/modern-agile-team/9term-main-back/pull/52#discussion_r2189004997

rrgks6221 commented on src/rooms/dtos/requests/update-room.dto.ts:
```diff
@@ -1,4 +1,75 @@
+  mode: RoomMode;
+
+  @IsInt()
+  @Type(() => Number)
```
UpdateRoomDto는 body로 값을 받고있죠??
그러면 transform 과정은 필요없을거같습니다

rrgks6221 commented on src/rooms/dtos/requests/update-room.dto.ts:
```diff
@@ -1,4 +1,75 @@
+
+  @IsInt()
+  @Type(() => Number)
+  @Min(1)
```
description을 봤을 때 현재는 3~6 사이의 값만 허용해야할거같아서 pipe에서 validation을 한다면 min(3), max(6)으로 하는게 좋을거같아요.
개인적인 의견으로는 solo인 경우, team인 경우에 따라 도메인 불변식이 들어가기 때문에 pipe에서는 int형태 검사만 하고 엔티티에서 검증하는게 낫다고 생각하긴 해요. 도메인 레이어(엔티티)에서는 해당 유효성 검사를 꼭 해야하는데 pipe에서도 하면 도메인 규칙이 여러곳에 퍼짐

yxxjxxeee replied:
말씀하신 내용 이해했습니다. 제 생각에도 엔티티에서만 검증하는 게 좋을 것 같아요!
pipe에서는 int 타입 검사만 하고 엔티티에서 유효성 검사를 하는 방식으로 반영하겠습니다.

rrgks6221 commented on src/rooms/dtos/requests/update-room.dto.ts:
```diff
@@ -1,4 +1,75 @@
+    if (value === 'true') return true;
+    if (value === 'false') return false;
+    return value;
+  })
```
마찬가지로 body로 받는 값이라면 변환이 필요없을거같아요.
아니면 string도 허용하는 것일까요?

yxxjxxeee replied:
제가 착각했습니다! query나 param은 문자열로 들어와서 transform이 필요한데 body에서도 해줘야 하는 줄 알았어요.

rrgks6221 commented on src/rooms/dtos/responses/pagination-room-response.dto.ts:
```diff
@@ -0,0 +1,19 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { ResRoomDto } from './room-response.dto';
+
+export class PaginationRoomResponseDto {
```
현재도 문제는 없습니다만
페이징 처리된 room이라는 의미로 `paginatedRoomResponseDto` 는 어떻게 생각하시나요??

yxxjxxeee replied:
더 명확해서 좋은 것 같습니다!

rrgks6221 commented on src/common/pagination/paginate.util.ts:
```diff
@@ -0,0 +1,15 @@
+export function paginate<T extends { id: number | null }>(
```
지금 이 함수를 service에서 사용하고 있는거같은데 런타임에서 엔티티를 받기 때문에 id가 null인 경우가 없지 않나요??

yxxjxxeee replied:
Room 엔티티가 생성 시점에 id: null을 가지므로 타입을 number | null로 설정했습니다.
paginate는 실제로는 항상 저장된 엔티티(id가 number)를 받지만
Room[] 타입과 호환되려면 null을 허용해야 타입 에러가 없어서 그랬습니다. 
혹시 엔티티에서 생성 전, DB 저장 후 타입을 분리하는 게 나을까요?

```typescript
export class Room {
  private constructor(
    public id: number | null,
    public name: string,
    public mode: RoomMode,
    public maxPlayers: number,
    public currentPlayers: number,
    public isPrivate: boolean,
    private _hashedPassword: string | null,
    private _status: RoomStatus,
  ) {
    this.validate();
  }

  static create(
    name: string,
    mode: RoomMode,
    maxPlayers: number,
    isPrivate: boolean,
    hashedPassword: string | null,
  ): Room {
    return new Room(
      null,
      name,
      mode,
      maxPlayers,
      ROOM_CONSTANTS.INITIAL_CURRENT_PLAYERS,
      isPrivate,
      hashedPassword,
      RoomStatus.WAITING,
    );
  }
```

rrgks6221 replied:
아 고유 식별자가 데이터베이스를 통해서 생성되는 구조라 생성 시 id null을 허용해서 그렇군요
https://github.com/modern-agile-team/10term-pollycasso-back/pull/25#discussion_r2462668529 이 코멘트를 반영하신다면 리포지토리 레이어에서 해당 유틸함수를 실행시킬테니 인풋 타입으로 null을 허용하지 않아도 문제가 없을거같은데 맞을까요??

1. 리포지토리에서 prisma 스키마와 일치하는 객체를 파싱
2. paginate 함수를 통해 paging 된 prisma 객체를 생성
3. 엔티티로 변환된 인터페이스 응답
4. 서비스에서 파싱

yxxjxxeee replied:
그러네요!! 레포지토리에서 DB에서 가져온 Room 엔티티를 대상으로 paginate를 사용하면 문제가 해결되겠네요..

rrgks6221 commented on src/rooms/rooms.service.ts:
```diff
@@ -44,22 +44,25 @@ export class RoomsService {
+    query: QueryRoomDto,
+  ): Promise<{ rooms: Room[]; hasNextPage: boolean; nextCursor: number | null }> {
+    const rooms = await this.roomsRepository.findAllRooms(query, ROOM_CONSTANTS.ROOMS_PER_PAGE);
+    const { data, hasNextPage, nextCursor } = paginate(rooms, ROOM_CONSTANTS.ROOMS_PER_PAGE);
```
paginate 구현을 보면 아래와 같은거로 파악했는데요

- 다음 커서 생성
- 다음 페이지 존재 여부 확인
- 다음 페이지 존재 여부 확인을 위해 1개 더 가져온 데이터를 제거

아래 같은 부분을 통해서 service보다는 repository에서 하는게 맞다고 생각하는데 어떻게 생각하시나요??

- 다음 페이지 존재여부를 service에서 계산해야하기 때문에 repository에서 limit에 해당하는 데이터보다 1개 더 응답하는 문제
- 페이지네이션 자체가 백엔드 내에서도 DB 의존적인 부분
- 커서라는 값 자체도 repository에서 필요한 정보이기 때문에 생성을 repository에서 담당하는게 더 좋다고 생각함

yxxjxxeee replied:
좋은 것 같습니다! 페이지네이션 계산이 DB 쿼리랑 붙어 있어서 repository에서 처리하는 게 맞는 것 같아요.
그래서 repository에서 paginate()를 호출하도록 변경했습니다!

rrgks6221 reviewed (commented):
지금 커서를 보면 number 형태로 그대로 받고있는것으로 파악했는데요
일반적으로 cursor는 인코딩하여 string 형태로 많이 사용합니다. 보통 base64로 많이 인코딩하여 사용해요
당장에 체감되지 않는 내용이고 이미 클라이언트에서 해당 기능을 사용중일테니 변경하기는 어려울 수 있지만 참고하여 나중에 적용하시거나 공부해보시면 좋을거같아요.

이유는

1. 클라이언트에게 백엔드 구현을 숨김
2. 커서의 소스가 변경되는 경우 타입변형이 일어날 수 있고 도메인에 따라 커서 소스가 다른 경우 일관되지 않은 인터페이스를 제공하게 됨
   1. room에서는 id를 커서로 사용함, member에서는 updatedAt을 사용함 
4. 복합 커서가 사용될 경우(아마 10기분들 프로젝트에서는 이러한 케이스는 없읅같긴 함) 객체를 파싱해야하는 문제가 생김
   1. id로 생성순 정렬이 부족한 경우(id를 uuid로 사용한다거나) {createdAt: Date, id: string} 를 커서로 사용해야하는 경우가 있음
   2. 다중정렬이 제공될 경우 복합 커서가 필요함

yxxjxxeee committed:
refactor: (#20) paginate 처리 및 validation, swagger 개선

yxxjxxeee committed:
refactor: (#20) 기존 개별 필드를 props 객체로 통합

kimjiwonpg98 commented on src/rooms/entities/rooms.entity.ts:
```diff
@@ -14,66 +32,63 @@ interface RoomUpdateParams {
+    if (props.maxPlayers !== undefined) this.props.maxPlayers = props.maxPlayers;
+    if (props.status !== undefined) this.props.status = props.status;
+    if (props.isPrivate !== undefined) this.props.isPrivate = props.isPrivate;
+    if (props.hashedPassword !== undefined) this.props.hashedPassword = props.hashedPassword;
```
```suggestion
    const updates = Object.fromEntries(
        Object.entries(props).filter(([, value]) => value !== undefined)
    );

Object.assign(this.props, updates);
```

kimjiwonpg98 replied:
props가 추가되도 변경하지 않도록 이런식으로 리펙토링하는건 어떨까요?

yxxjxxeee replied:
확실히 유지보수가 편해질 것 같아서 좋은 것 같습니다!

yxxjxxeee committed:
refactor: (#20) repository로 paginate 이동 및 props 업데이트 방식 개선

- service에서 처리하던 paginate 로직을 repository로 이동
- if문 대신 filter를 사용해 props 업데이트 간소화
